sky/cloud_stores.py

```diff
@@ -189,30 +194,84 @@ def is_directory(self, url: str) -> bool:
-        # If there's more, we'd need to check if it's a directory or a file.
-        container_client = data_utils.create_az_client(
-            client_type='container', container_url=container_url)
-        num_objects = 0
-        for blob in container_client.list_blobs(name_starts_with=path):
-            if blob.name == path:
-                return False
-            num_objects += 1
-            if num_objects > 1:
-                return True
-
-        # A directory with few or no items
-        return True
+        resource_group_name = azure.get_az_resource_group(storage_account_name)
+        role_assignment_start = time.time()
+        refresh_client = False
+        role_assigned = False
+
+        while (time.time() - role_assignment_start <
+               constants.WAIT_FOR_STORAGE_ACCOUNT_ROLE_ASSIGNMENT):
+            container_client = data_utils.create_az_client(
+                client_type='container',
+                container_url=container_url,
+                storage_account_name=storage_account_name,
+                resource_group_name=resource_group_name,
+                refresh_client=refresh_client)
+
+            num_objects = 0
+            try:
+                for blob in container_client.list_blobs(name_starts_with=path):
+                    if blob.name == path:
+                        return False
+                    num_objects += 1
+                    if num_objects > 1:
+                        return True
+            except azure.exceptions().HttpResponseError as e:
+                # Handle case where user lacks sufficient IAM role for
+                # a private container in the same subscription. Attempt to
+                # assign appropriate role to current user.
+                if 'AuthorizationPermissionMismatch' in str(e):
+                    if not role_assigned:
+                        logger.info('Failed to list blobs in container '
+                                    f'{container_url!r}. This implies '
+                                    'insufficient IAM role for storage account'
+                                    f' {storage_account_name!r}.')
+                        azure.assign_storage_account_iam_role(
+                            storage_account_name=storage_account_name,
+                            resource_group_name=resource_group_name)
+                        role_assigned = True
+                        refresh_client = True
+                    else:
+                        logger.info(
+                            'Waiting due to the propagation delay of IAM '
+                            'role assignment to the storage account '
+                            f'{storage_account_name!r}.')
+                        time.sleep(
+                            constants.RETRY_INTERVAL_AFTER_ROLE_ASSIGNMENT)
+                    continue
+                raise
+            # A directory with few or no items
+            return True
```

Comment:
This is what I tried at first, but the error is raised when we iterate through the object within the for-loop `for blob in blobs:` rather than when the object is obtained with `blobs = container_client.list_blobs(name_starts_with=path)`. So the suggested code won't be able to catch the error.